Add ArgsEscaped support for CRI

kiashok opened the pull request:
This PR addresses windows images that are broken when created using the "shell form" entrypoint/cmd in Dockerfile. Associated bug - https://github.com/containerd/containerd/issues/5067
The approach taken to fix this bug is as discussed by @kevpar here - https://github.com/containerd/containerd/pull/6479#discussion_r813320420 

Docker uses non-OCI field called ArgsEscaped to address how the image 'ENTRYPOINT' and 'CMD' are interpreted on windows. This PR adds support for ArgsEscaped in containerd without changes to the OCI spec and also adds changes to handle cases when image ENTRYPOINT and CMD can be overriden by 'Command' and 'Args' specified in the podspec.

**The fix in this PR does the following:**
- Reads the ArgsEscaped value from the image got from docker without making OCI spec changes (In file image.go)
- Image ENTRYPOINT and CMD fields can be overriden by the container 'Command' and 'Args' fields specified through the podspec. But the ArgsEscaped field should really only affect the behavior of the image values, and not the overall evaluation of args for the container. Therefore, we need to evaluate if the first arg in our list is coming from image entrypoint/cmd and **only if** it is coming from the image, we should we treat the first arg as already escaped.  (in file spec.go spec_windows.go)

Regression tests have also been added to this PR.

jterry75 commented:
I dont do any CRI plugin work anymore so I will defer to @kevpar / @dcantah here. But thank you for picking up this work. It would be nice to see CRI support for ArgsEscaped images working on Kube+CriContainerd

dcantah commented on oci/spec_opts.go:
```diff
@@ -232,6 +232,16 @@ func WithProcessArgs(args ...string) SpecOpts {
 	}
 }
 
+// WithProcessCommandLine replaces the command line on the generated spec
```
Would be good to call out that Args on the spec will get zero'd from this

dcantah commented on pkg/cri/opts/spec_windows.go:
```diff
@@ -274,3 +276,40 @@ func WithDevices(config *runtime.ContainerConfig) oci.SpecOpts {
+		// entrypoint and cmd should be interpreted.
+		//
+		args, firstArgFromImg, err := getArgs(image.Entrypoint, image.Cmd, config.GetCommand(), config.GetArgs())
+
```
Newline between error check

dcantah commented on pkg/cri/opts/spec_windows.go:
```diff
@@ -274,3 +276,40 @@ func WithDevices(config *runtime.ContainerConfig) oci.SpecOpts {
+		}
+
+		var cmdLine string
+
```
Also odd newline

dcantah commented on pkg/cri/server/container_create_linux_test.go:
```diff
@@ -113,11 +114,17 @@ func getCreateContainerTestData() (*runtime.ContainerConfig, *runtime.PodSandbox
 		Entrypoint: []string{"/entrypoint"},
 		Cmd:        []string{"cmd"},
 		WorkingDir: "/workspace",
+		User:       "ContainerUser",
```
Do we now need this for linux?

kiashok replied:
No we don't need this for linux since the interpretation of args is different

dcantah replied:
No I meant the User addition of `ContainerUser` that didn't used to be on the spec here

dcantah commented on pkg/cri/server/container_create_windows_test.go:
```diff
@@ -245,3 +256,211 @@ func TestHostProcessRequirements(t *testing.T) {
+	testPid := uint32(1234)
+	nsPath := "test-cni"
+	c := newTestCRIService()
+	ImgEntrypoint := []string{"C:\\My Folder\\MyProcess.exe", "-arg1", "test value"}
```
Can use backtick strings to avoid double backslash

```go
path := `C:\path\to\thing`
```

dcantah replied:
```suggestion
	imgEntrypoint := []string{"C:\\My Folder\\MyProcess.exe", "-arg1", "test value"}
```

dcantah commented on pkg/cri/server/container_create_windows_test.go:
```diff
@@ -245,3 +256,211 @@ func TestHostProcessRequirements(t *testing.T) {
+	nsPath := "test-cni"
+	c := newTestCRIService()
+	ImgEntrypoint := []string{"C:\\My Folder\\MyProcess.exe", "-arg1", "test value"}
+	ImgCmd := []string{"cmd", "-args", "hello world"}
```
```suggestion
	imgCmd := []string{"cmd", "-args", "hello world"}
```

dcantah commented on pkg/cri/store/image/image.go:
```diff
@@ -134,17 +141,38 @@ func getImage(ctx context.Context, i containerd.Image) (*Image, error) {
+	// Deserialize the extended image format for Windows.
+	var ociImageExtended imageExtended
+	if err := json.Unmarshal(rb, &ociImageExtended); err != nil {
+		return nil, errors.Wrapf(err, "unmarshal image config for ArgsEscaped %s", rb)
```
pkg/errors was removed as a dependency in this repo so best not to reintroduce it. Can use the built in wrapping support in Go.
```suggestion
		return nil, fmt.Errorf("unmarshal image config for ArgsEscaped %s: %w", rb, err)
```

dcantah commented on pkg/cri/store/image/image.go:
```diff
@@ -134,17 +141,38 @@ func getImage(ctx context.Context, i containerd.Image) (*Image, error) {
-		return nil, fmt.Errorf("failed to get OCI image spec: %w", err)
+		return nil, fmt.Errorf("read image config from content store: %w", err)
 	}
+	var ociimage imagespec.Image
```
```suggestion
	var ociImage imagespec.Image
```

dcantah commented on pkg/cri/store/image/image.go:
```diff
@@ -134,17 +141,38 @@ func getImage(ctx context.Context, i containerd.Image) (*Image, error) {
+	rb, err := content.ReadBlob(ctx, i.ContentStore(), desc)
 	if err != nil {
-		return nil, fmt.Errorf("failed to get OCI image spec: %w", err)
+		return nil, fmt.Errorf("read image config from content store: %w", err)
```
I'd probably put "failed to read image config from content store"

dcantah commented on pkg/cri/store/image/image.go:
```diff
@@ -134,17 +141,38 @@ func getImage(ctx context.Context, i containerd.Image) (*Image, error) {
 	id := desc.Digest.String()
 
-	spec, err := i.Spec(ctx)
+	rb, err := content.ReadBlob(ctx, i.ContentStore(), desc)
```
A more descriptive name to signify what the blob we just read is

kiashok replied:
Does contentStoreImgBlob sound ok?

dcantah commented:
I really think for full testing sanity we should host an image that is built with docker that would set ArgsEscaped, and then have a test in the CRI /integration suite that launches a container and verifies we're all good.

dcantah commented:
I think the Project Checks are failing because you're using github.com/pkg/errors in a spot here and it was removed as a direct dependency

dcantah commented:
PR description is great but please add some description to the commit message as well describing the need for this change and what it fixes

dcantah commented on pkg/cri/opts/spec.go:
```diff
@@ -51,27 +51,51 @@ func WithoutRoot(ctx context.Context, client oci.Client, c *containers.Container
+	// Entrypoint or Cmd. If the first arg instead comes from the container config (e.g. overriding the image values),
+	// it should be false.
+	// Essentially this means firstArgFromImg should be true iff:
+	// - Container entrypoint is nil and image entrypoint is not empty
+	// - Container entrypoint is nil and image entrypoint is empty and container cmd is empty and image cmd is not empty
+	// - Container entrypoint is empty and container cmd is empty and image cmd is not empty
```
A table would help here (although hard to do in comments..), or commas to help break em up a little

dcantah commented on pkg/cri/opts/spec.go:
```diff
@@ -51,27 +51,51 @@ func WithoutRoot(ctx context.Context, client oci.Client, c *containers.Container
-			// Copy array to avoid data race.
-			if len(args) == 0 {
-				args = append([]string{}, image.Cmd...)
+func getArgs(imgEntrypoint []string, imgCmd []string, ctrEntrypoint []string, ctrCmd []string) ([]string, bool, error) {
```
Even though it's not exported, this should probably have a comment describing what's going on in here as it's non-trivial

mikebrow reviewed (commented):
Thank you! I see a lot of work here.  IMO this change may best be done after the merge of https://github.com/opencontainers/image-spec/pull/892 till then this could proto on a branch with that change merged, which would mean a lot less code necessary, wouldn't have to revert some of Maksym's recent changes that moved get spec to the client.. wouldn't to unmarshal twice to get the extended (deprecated docker) field....

dcantah commented:
I think the reason was that PR had been stalled for awhile and wasn't sure if it'd be done in time for 1.7. Would it be undesirable to use the double unmarshal to get the ArgsEscaped field in the meantime until we have that in imagespec, and we can do a follow-up change? 

For the reverting Maksym's changes bit, which part of this change was that?

dcantah commented:
@mikebrow pingaroo on the above. Question: given we don't vendor an exact tag of the spec here today already, am I right in assuming we could re-vendor image-spec from the latest in main once that change is merged? If it's highly preferred to wait for that to land we can try and drive that forward ASAP. If that doesn't seem to be making progress in time for 1.7, could we move forward with what's here for now and retrofit after?

marosset commented:
https://github.com/opencontainers/image-spec/pull/892 merged :)

jterry75 commented:
RE: https://github.com/containerd/containerd/pull/7483 to update containerd to the spec for non CRI

estesp commented:
Seems this patch must introduce a logic error with config env vars; all CI runs are now failing the test `TestContainerSpecDefaultPath`

jterry75 commented on integration/windows_hostprocess_test.go:
```diff
@@ -124,3 +124,35 @@ func runHostProcess(t *testing.T, expectErr bool, image string, action hpcAction
+	cnConfig := ContainerConfig(
+		containerName,
+		testImage,
+		WithCommand("ping", "-t", "127.0.0.1"),
```
Should we have a default test case with no cmd merge?

kiashok replied:
Done

jterry75 commented on oci/spec_opts.go:
```diff
@@ -231,6 +231,16 @@ func WithProcessArgs(args ...string) SpecOpts {
 }
 
+// WithProcessCommandLine replaces the command line on the generated spec
+func WithProcessCommandLine(cmdLine string) SpecOpts {
```
This only makes sense for Windows. I think we should set it in the oci_win file itself.

jterry75 commented on pkg/cri/opts/spec.go:
```diff
@@ -51,27 +51,54 @@ func WithoutRoot(ctx context.Context, client oci.Client, c *containers.Container
-					command = append([]string{}, image.Entrypoint...)
-				}
+		}
+		if entrypoint == nil {
```
This seems really odd to me. Why would a zero length be treated differently than nil here. IE: if len(entrypoint)== 0 and entrypoint != nil why would we not want the image entrypoint?

jterry75 commented on pkg/cri/opts/spec.go:
```diff
@@ -51,27 +51,54 @@ func WithoutRoot(ctx context.Context, client oci.Client, c *containers.Container
+// WithProcessArgs sets the process args on the spec based on the image and runtime config
+func WithProcessArgs(config *runtime.ContainerConfig, image *imagespec.ImageConfig) oci.SpecOpts {
+	return func(ctx context.Context, client oci.Client, c *containers.Container, s *runtimespec.Spec) (err error) {
+		args, _, err := getArgs(image.Entrypoint, image.Cmd, config.GetCommand(), config.GetArgs())
```
This doesnt feel right. If `firstArgFromImg` is false, then we dont want `WithProcessArgs` right which would append the args to the image. It really feels like the logic should be:

If entrypointFromConfig append argsFromConfig OR argsFromImage
if entrypointFromImage append argsFromConfig OR argsFromImage

Right?

kiashok replied:
Not sure I understand this - Only if firstArgsFromIImg is set, we will escaped and append to CommandLine. container_create_windows.go and WithProcessCommandLineOrArgs() function have more details.

jterry75 commented on pkg/cri/server/container_create_windows.go:
```diff
@@ -46,12 +47,16 @@ func (c *criService) containerSpec(
-		customopts.WithProcessArgs(config, imageConfig),
+	imageConfig := &image.ImageSpec.Config
+	var specOpts []oci.SpecOpts
+	if image.ArgsEscaped {
```
I see why you are doing this now. It really feels to me that you dont know this at this level and you should just call "merge" on it regardless. The "merge" code should determine to set Args vs CommandLine. IE: If you have both entrypoint and cmd from config, then the image ArgsEscaped doesnt matter at all. So we should handle it at the lower layer

kiashok replied:
I have moved the check of argsEscaped to a lower level. Also removed a bunch of other unnecessary changes that were in the PR after OCI changes

kiashok replied:
I have moved the check of argsEscaped to a lower level. Also removed a bunch of other unnecessary changes that were in the PR after OCI changes

jterry75 reviewed (changes requested):
Great start. Left some comments.

kiashok commented:
> Great start. Left some comments.

Hi @jterry75 , I addresses the comments. Could you please let me know if they look ok? Thanks a lot!

kiashok commented:
Hi @dcantah , would you be able to take a second look at the changes when you have some time please? Thanks!

dcantah commented:
@kiashok Yep! I'll try and get to this tomorrow evening

kiashok commented:
> @kiashok Yep! I'll try and get to this tomorrow evening

Thanks Danny!

kiashok commented:
> Today's lookin a little slammed.. 🫠 I'll definitely get to this sometime this week @kiashok (please ping if I don't)

Sure Danny no worries, thanks a lot! :)

dcantah commented on integration/images/image_list.go:
```diff
@@ -30,12 +30,13 @@ var imageListFile = flag.String("image-list", "", "The TOML file containing the
+	ResourceConsumer     string
+	VolumeCopyUp         string
+	VolumeOwnership      string
+	ArgsEscapedTestImage string
```
No need for TestImage in the field name as we know that these are all images being used for testing

dcantah replied:
```suggestion
	ArgsEscaped string
```

dcantah commented on integration/windows_hostprocess_test.go:
```diff
@@ -124,3 +124,47 @@ func runHostProcess(t *testing.T, expectErr bool, image string, action hpcAction
+
+	t.Log("Start the container")
+	_, err = t, runtimeService.StartContainer(cn)
+
```
remove empty line

dcantah replied:
Might as well require.NoError here though

dcantah commented on integration/windows_hostprocess_test.go:
```diff
@@ -124,3 +124,47 @@ func runHostProcess(t *testing.T, expectErr bool, image string, action hpcAction
+	require.NoError(t, runtimeService.RemoveContainer(cn))
+}
+
+func TestArgsEscapedImagesOnWindows(t *testing.T) {
```
Is there any reason we're testing with a hostprocess container instead of regular? Not that I think my command line parsing code isn't rock solid 🤣 but feel it'd be more useful to test against HCS were most of the bug reports had been for this.

dcantah commented on pkg/cri/server/container_create_windows_test.go:
```diff
@@ -245,3 +250,179 @@ func TestHostProcessRequirements(t *testing.T) {
+	testPid := uint32(1234)
+	nsPath := "test-cni"
+	c := newTestCRIService()
+	ImgEntrypoint := []string{`"C:\My Folder\MyProcess.exe" -arg1 "test value"`}
+	ImgCmd := []string{`cmd -args "hello world"`}
+	args := []string{`cmd -args "additional args"`}
+
+	expectedCommandLine := `"C:\My Folder\MyProcess.exe" -arg1 "test value" "cmd -args \"additional args\""`
+
+	containerConfig, sandboxConfig, imageConfig := getCreateContainerTestDataForArgsEscapedTests(ImgEntrypoint, ImgCmd, nil, args)
+	runtimeSpec, err := c.containerSpec(testID, testSandboxID, testPid, nsPath, testContainerName, testImageName, containerConfig, sandboxConfig, imageConfig, nil, config.Runtime{})
+	assert.NoError(t, err)
+	assert.NotNil(t, runtimeSpec)
+
+	checkRuntimeSpec(t, runtimeSpec, nil, expectedCommandLine)
+}
+
+// check image entrypoint and cmd in exec form without overriding with container command and args and verify expected runtime spec
+func TestExecFormImgEntrypointCmdWithoutCtrArgs(t *testing.T) {
+	testID := "test-id"
+	testSandboxID := "sandbox-id"
+	testContainerName := "container-name"
+	testPid := uint32(1234)
+	nsPath := "test-cni"
+	c := newTestCRIService()
+	imgEntrypoint := []string{`C:\My Folder\MyProcess.exe`, "-arg1", "test value"}
+	imgCmd := []string{"cmd", "-args", "hello world"}
+
+	expectedArgs := []string{`C:\My Folder\MyProcess.exe`, "-arg1", "test value", "cmd", "-args", "hello world"}
+
+	containerConfig, sandboxConfig, imageConfig := getCreateContainerTestDataForArgsEscapedTests(imgEntrypoint, imgCmd, nil, nil)
+	runtimeSpec, err := c.containerSpec(testID, testSandboxID, testPid, nsPath, testContainerName, testImageName, containerConfig, sandboxConfig, imageConfig, nil, config.Runtime{})
+	assert.NoError(t, err)
+	assert.NotNil(t, runtimeSpec)
+
+	checkRuntimeSpec(t, runtimeSpec, expectedArgs, "")
+}
+
+// override image cmd by container args in exec form and verify expected runtime spec
+func TestExecFormImgEntrypointCmdWithCtrArgs(t *testing.T) {
```
I was trying to think of how to simplify these names, but that led to thinking that these should probably just be a single test where these are testcases and use t.Run(). So TestArgsEscaped (or any name you think makes sense here) and then something like https://github.com/containerd/containerd/blob/main/pkg/cri/sbserver/container_stats_list_linux_test.go#L63-L89 (arbitrarily chosen, there's a ton of t.Run() examples to choose from here)

dcantah commented on pkg/cri/server/container_create_windows_test.go:
```diff
@@ -245,3 +250,179 @@ func TestHostProcessRequirements(t *testing.T) {
+	return nil
+}
+
+func getCreateContainerTestDataForArgsEscapedTests(imgEntrypoint []string, imgCmd []string, command []string, args []string) (*runtime.ContainerConfig, *runtime.PodSandboxConfig, *imagespec.ImageConfig) {
```
If the combining all these tests into one suggestion is taken we should probably just use getCreateContainerTestData and edit the imageConfig in the one test

dcantah commented:
@kiashok I'll try and get a final pass sometime this week. Thanks for looking at the feedback!

kiashok commented:
> @kiashok I'll try and get a final pass sometime this week. Thanks for looking at the feedback!



> @kiashok I'll try and get a final pass sometime this week. Thanks for looking at the feedback!

Thanks a lot for looking into this @dcantah

mxpv commented on pkg/cri/opts/spec_windows_opts.go:
```diff
@@ -1,3 +1,6 @@
+//go:build windows
```
Do we need this build constraints here?
We should be able to generate OCI cross platform.
See https://github.com/containerd/containerd/pull/7928 for context.

kiashok replied:
I had tried to avoid adding the windows specific builds before the last update as well but the problem is with importing windows pkg ( the use of windows.EscapeArg() call in escapeAndCombineArgsWindows() ). 
There seems to be a restriction :

package github.com/containerd/containerd/cmd/containerd
        imports github.com/containerd/containerd/cmd/containerd/builtins
        imports github.com/containerd/containerd/pkg/cri
        imports github.com/containerd/containerd/pkg/cri/sbserver
        imports github.com/containerd/containerd/pkg/cri/opts
        **imports golang.org/x/sys/windows: build constraints exclude all Go files in C:\Users\kiashok\go\src\github.com\containerd\upstreamCtrd\argsEscaped\containerd\vendor\golang.org\x\sys\windows**

In the example Danny pointed out too, there is a build restriction like "//go:build !linux && !windows"

kiashok replied:
ping on this :)

mxpv replied:
If it's platform dependent, then yes, we should move it to a `_windows.go` file.

kiashok replied:
Alright, will move windows specific functions from this file to pkg/cri/opts/spec_windows.go

mxpv commented on pkg/cri/server/container_create_windows.go:
```diff
@@ -51,9 +51,8 @@ func (c *criService) containerSpec(
-		customopts.WithProcessArgs(config, imageConfig),
-	}
+	var specOpts []oci.SpecOpts
+	specOpts = append(specOpts, customopts.WithProcessCommandLineOrArgs(config, imageConfig))
```
Could you pls backport this to `sbserver` as well?

kiashok replied:
The code refactoring seems to be very different in sbserver. Do we need to put windows specific functions under _windows.go there too?

kiashok commented:
@mxpv could you please take a look at the new changes and see if they look good? 

Also, the CI failures seem to be unrelated to the change

dmcgowan commented on pkg/cri/opts/spec_nonwindows.go:
```diff
@@ -0,0 +1,29 @@
+)
+
+func WithProcessCommandLineOrArgs(config *runtime.ContainerConfig, image *imagespec.ImageConfig) oci.SpecOpts {
+	panic("not supported")
```
Just make it a no-opt then?

kiashok replied:
Not sure I understand. What do you mean when you say no-opt? Just return nil instead of panic?

kiashok replied:
@dmcgowan ping on this :)

dmcgowan replied:
Sorry, meant to say no-op. Just return a SpecOpt which doesn't do anything or returns `errdefs.ErrNotImplemented`

dmcgowan commented on pkg/cri/sbserver/container_create.go:
```diff
@@ -730,9 +730,8 @@ func (c *criService) buildWindowsSpec(
-		customopts.WithProcessArgs(config, imageConfig),
-	}
+	var specOpts []oci.SpecOpts
+	specOpts = append(specOpts, customopts.WithProcessCommandLineOrArgs(config, imageConfig))
```
If the function is only intended for use on Windows, then it would be better to add Windows to the name of the function.

kiashok replied:
I could change this but CommandLine itself is a windows only spec right - https://github.com/opencontainers/runtime-spec/blob/main/specs-go/config.go#L46 ? Doe WithProcessCommandLineOrArgsForWindows() sound better?

kiashok replied:
@dmcgowan ping on this :)

dmcgowan replied:
Renaming sounds better if it is truly something only for Windows. Removing the panic would be good as well, we don't normally have panics in the runtime path to represent not implemented,  unless its on startup.

dmcgowan commented on pkg/cri/opts/spec_windows.go:
```diff
@@ -0,0 +1,121 @@
+	runtime "k8s.io/cri-api/pkg/apis/runtime/v1"
+)
+
+func escapeAndCombineArgsWindows(args []string) string {
```
If this is the only function which has Windows specific logic, seems better to only have this in the windows only part of the build. The rest of the logic may be intended for windows, but with sbserver the logic does not need to be only compiled on Windows.

kiashok replied:
escapeAndCombineArgsWindows() is called from WithProcessCommandLineOrArgs() which is the function below this, making it windows only. And the other function in this file getArgs() is a helper function called only from WithProcessCommandLineOrArgs() so it made more sense to have them all as part of the spec_windows.go file.

kiashok replied:
@dmcgowan ping on this :)

kiashok replied:
renamed WithProcessCommandLineOrArgs() function in this file to windows only - WithProcessCommandLineOrArgsForWindows(). the getArgs() funciton is the only other one in this file and it is a helper function called from WithProcessCommandLineOrArgsForWindows() only. Therefore leaving all these functions to this file as that seems to make more sense.

kiashok commented:
Looks like accidentally deleted the fork that had the branch with the changes in this PR. Addressed the commentd @dmcgowan left and wanted to submit new changes and realized this. Closing this PR as I created a new one here - https://github.com/containerd/containerd/pull/8198. Apologies for this!
